plugins/org.eclipse.n4js/src/org/eclipse/n4js/typesystem/DefaultUnsupportedExpressionTypeHelper.java

```diff
@@ -12,25 +12,83 @@
 
 import org.eclipse.emf.ecore.EObject;
 import org.eclipse.n4js.n4JS.Expression;
+import org.eclipse.n4js.n4JS.JSXElement;
+import org.eclipse.n4js.n4JS.JSXPropertyAttribute;
+import org.eclipse.n4js.n4jsx.helpers.ReactHelper;
+import org.eclipse.n4js.ts.typeRefs.ParameterizedTypeRef;
 import org.eclipse.n4js.ts.typeRefs.TypeRef;
+import org.eclipse.n4js.ts.types.TClassifier;
+import org.eclipse.n4js.ts.utils.TypeUtils;
 import org.eclipse.xsemantics.runtime.RuleEnvironment;
 
+import com.google.inject.Inject;
+
 /**
  * Default implementation of {@link UnsupportedExpressionTypeHelper} that always throws an
  * {@link UnsupportedOperationException}. In N4JS, the method {@link #typeExpression(Expression, RuleEnvironment)} must
  * never be called as all types of expressions should be handled in Xsemantics.
  */
 public class DefaultUnsupportedExpressionTypeHelper implements UnsupportedExpressionTypeHelper {
+	@Inject
+	ReactHelper reactLookupHelper;
+	@Inject
+	N4JSTypeSystem ts;
+
+	/**
+	 * In default case throws {@link UnsupportedOperationException}.
+	 * <p>
+	 * When provided @ {@code expression} is link JSXElement} tries to find proper {@link TypeRef}, that is to return
+	 * the type of JSX element as React.Element
+	 *
+	 * @param expression
+	 *            the JSX element
+	 * @param G
+	 *            the rule environment
+	 *
+	 * @return the type ref to React.Element
+	 * @throws UnsupportedOperationException
+	 *             when expression can not be typed
+	 */
 	@Override
 	public TypeRef typeExpression(Expression expression, RuleEnvironment G) {
+		if (expression instanceof JSXElement) {
+			TClassifier classifierReactElement = reactLookupHelper.lookUpReactElement(expression);
+			if (classifierReactElement != null) {
+				ParameterizedTypeRef typeRef = TypeUtils.createTypeRef(classifierReactElement);
+				return typeRef;
+			}
+		}
```

Comment:
Interface `UnsupportedExpressionTypeHelper` and all implementing classes were only a work-around required because N4JSX was a separate Xtext language. So I propose to remove that interface, all implementing classes, and all related code. The above type inference rule should then be added to file `n4js.xsemantics`.

Reply:
done.